ddtrace/ci_visibility/_recorder.py

```diff
@@ -0,0 +1,143 @@
+
+
+class CIVisibility(Service):
+    _instance = None  # type: Optional[CIVisibility]
+    enabled = False
+
+    def __init__(self, tracer=None, config=None, service=None):
+        # type: (Optional[Tracer], Optional[IntegrationConfig], Optional[str]) -> None
+        super(CIVisibility, self).__init__()
+
+        self.tracer = tracer or ddtrace.tracer
+        self.config = config  # type: Optional[IntegrationConfig]
+        self._tags = ci.tags()  # type: Dict[str, str]
+        self._service = service
+        self._codeowners = None
+
+        int_service = None
+        if self.config is not None:
+            int_service = trace_utils.int_service(None, self.config)
+        # check if repository URL detected from environment or .git, and service name unchanged
+        if self._tags.get(ci.git.REPOSITORY_URL, None) and self.config and int_service == self.config._default_service:
+            self._service = _extract_repository_name_from_url(self._tags[ci.git.REPOSITORY_URL])
+        elif self._service is None and int_service is not None:
+            self._service = int_service
+
+        try:
+            from ddtrace.internal.codeowners import Codeowners
+
+            self._codeowners = Codeowners()
+        except ValueError:
+            log.warning("CODEOWNERS file is not available")
+        except Exception:
+            log.warning("Failed to load CODEOWNERS", exc_info=True)
+
+    @classmethod
+    def enable(cls, tracer=None, config=None, service=None):
+        # type: (Optional[Tracer], Optional[Any], Optional[str]) -> None
+
+        if cls._instance is not None:
+            log.debug("%s already enabled", cls.__name__)
+            return
+        log.debug("Enabling %s", cls.__name__)
+
+        cls._instance = recorder = cls(tracer=tracer, config=config, service=service)
+        cls.enabled = True
+
+        recorder.start()
+        forksafe.register(cls._restart)
+        atexit.register(cls.disable)
+
+        log.debug("%s enabled", cls.__name__)
+
+    @classmethod
+    def disable(cls):
+        # type: () -> None
+        if cls._instance is None:
+            log.debug("%s not enabled", cls.__name__)
+            return
+        log.debug("Disabling %s", cls.__name__)
+
+        forksafe.unregister(cls._restart)
+        atexit.unregister(cls.disable)
+
+        cls._instance.stop()
+        cls._instance = None
+        cls.enabled = False
+
+        log.debug("%s disabled", cls.__name__)
+
+    def _start_service(self):
+        # type: () -> None
+        tracer_filters = self.tracer._filters
+        if not any(isinstance(tracer_filter, TraceCiVisibilityFilter) for tracer_filter in tracer_filters):
+            tracer_filters += [TraceCiVisibilityFilter(self._tags, self._service)]  # type: ignore[arg-type]
+            self.tracer.configure(settings={"FILTERS": tracer_filters})
+
+    def _stop_service(self):
+        # type: () -> None
+        try:
+            self.tracer.shutdown()
```

Comment:
Do we need to shutdown the tracer here? We'll attempt to shutdown the tracer every time a process forks or shuts down. I'm not a big fan of this tight coupling between the CIVisibility service and the AgentWriter. Is there another workaround?

Reply:
I think we should investigate the implications of shutting the tracer down, but this discussion is out of scope of this PR as this is not new behavior but just extracted from the previous pytest plugin behavior.

FWIW, I personally don't think there's an issue with shutting down the tracer here as this will only happen on `pytest_sessionfinish()` which occurs once a test suite has been completed, i.e. the end of the job for CIVisibility (and therefore the tracer) purposes